admiral/pkg/clusters/shard_handler.go

```diff
@@ -208,12 +237,103 @@ func ConsumeIdentityConfigs(ctxLogger *log.Entry, ctx context.Context, configWri
-		//ctxLogger.Infof(common.CtxLogFormat, "UpdateShard", shard.Name, "", job.ClusterName, shard.Status.Conditions[0].Message)
-		//TODO: need to get updated shard crd spec and set status here
+	updatedShard := shard.DeepCopy()
+	updatedShard.Status.ClustersMonitored = len(shard.Spec.Clusters)
+	updatedShardStatusCondition := admiralapiv1.ShardStatusCondition{
+		Message:         "Shard handling complete",
+		Reason:          admiralapiv1.Processed,
+		Status:          admiralapiv1.TrueConditionStatus,
+		Type:            admiralapiv1.SyncComplete,
+		LastUpdatedTime: v1.Time{Time: time.Now()},
+	}
+	updatedShard.Status.FailureDetails = admiralapiv1.FailureDetails{
+		LastUpdatedTime: v1.Time{Time: time.Now()},
+		FailedClusters:  []admiralapiv1.FailedCluster{},
+	}
+	clusterFailedIdentitiesMap := make(map[string][]admiralapiv1.FailedIdentity)
+	for data := range results {
+		if data.Error != nil {
+			ctxLogger.Infof(common.CtxLogFormat, "ProcessResults", shard.Name, common.GetOperatorSyncNamespace(), data.ClusterName, data.Error.Error())
+			failedIdentityList := clusterFailedIdentitiesMap[data.ClusterName]
+			failedIdentity := admiralapiv1.FailedIdentity{
+				Name:    data.IdentityConfig.IdentityName,
+				Message: data.Error.Error(),
+			}
+			if failedIdentityList != nil {
+				clusterFailedIdentitiesMap[data.ClusterName] = slices.Insert(failedIdentityList, 0, failedIdentity)
+			} else {
+				clusterFailedIdentitiesMap[data.ClusterName] = []admiralapiv1.FailedIdentity{failedIdentity}
+			}
+			updatedShardStatusCondition.Reason = admiralapiv1.ErrorOccurred
+			updatedShardStatusCondition.Type = admiralapiv1.SyncFailed
+			updatedShardStatusCondition.LastUpdatedTime = v1.Time{Time: time.Now()}
+			updatedShard.Status.FailureDetails.LastUpdatedTime = v1.Time{Time: time.Now()}
+		}
+	}
+	for cluster, identities := range clusterFailedIdentitiesMap {
+		updatedShard.Status.FailureDetails.FailedClusters = append(updatedShard.Status.FailureDetails.FailedClusters, admiralapiv1.FailedCluster{
+			Name:             cluster,
+			FailedIdentities: identities,
+		})
+	}
+	//Just overwrite the first one because we already created a new condition entry in HandleEventForShard
+	updatedShard.Status.Conditions[0] = updatedShardStatusCondition
+	updatedShard.Status.LastUpdatedTime = v1.Time{Time: time.Now()}
+	err := updateShardStatus(ctx, ctxLogger, updatedShard, shard, cc)
+	if err != nil {
+		ctxLogger.Errorf(common.CtxLogFormat, "ProcessResults", shard.Name, shard.Namespace, "", err)
+	}
+	return err
+}
+
+func updateShardStatus(ctx context.Context, ctxLogger *log.Entry, obj *admiralapiv1.Shard, exist *admiralapiv1.Shard, cc admiralapi.Interface) error {
+	var err error
+	ctxLogger.Infof(common.CtxLogFormat, "UpdateShardStatus", obj.Namespace, "", "Updating Shard="+obj.Name)
+	exist, err = cc.AdmiralV1().Shards(obj.Namespace).Get(ctx, obj.Name, v1.GetOptions{})
```

Comment:
Why are we duplicating code ? Cant we just call the retry function

Reply:
Here I'm following the pattern from the other handlers, I think it's separated because we only want to retry if the err is the k8s.IsConflict error.